### ISSUE TITLE
Building Keycloak from source terminates with errors

### ISSUE THREAD

**Issue by christian-2** [keycloak/keycloak#20494]
### Before reporting an issue

- [X] I have searched existing issues
- [X] I have reproduced the issue with the latest release

### Area

dist/quarkus

### Describe the bug

I tried to build Keycloak by following instructions in [`building.md`](https://github.com/keycloak/keycloak/blob/main/docs/building.md) but ran into errors (see below).

### Version

21.1.1 (latest source code from Git)

### Expected behavior

I would have expected for Keycloak to build and its testsuite to execute successfully.

### Actual behavior

```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:3.0.0-M7:test (default-test) on project keycloak-quarkus-server-deployment: 
[ERROR] 
[ERROR] Please refer to /home/.../git/keycloak/quarkus/deployment/target/surefire-reports for the individual test results.
[ERROR] Please refer to dump files (if any exist) [date].dump, [date]-jvmRun[N].dump and [date].dumpstream.
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <args> -rf :keycloak-quarkus-server-deployment
```

The principal errors in `/home/.../git/keycloak/quarkus/deployment/target/surefire-reports` read: 

```
ERROR [org.infinispan.CONFIG] (keycloak-cache-init) ISPN000660: Cache org.infinispan.CONFIG start failed, stopping any running components: org.infinispan.commons.CacheConfigurationException: ISPN000659: Component org.infinispan.statetransfer.StateTransferManager failed to start
```

### How to Reproduce?

```
cat /etc/os-release # Debian 11
java -version # openjdk version "11.0.18" 2023-01-17
mvn -version # Apache Maven 3.6.3
git --version # git version 2.30.2
git clone https://github.com/keycloak/keycloak.git
cd keycloak
mvn clean install # fails w/ error as shown above
```

### Anything else?

I have previously raised this as discussion #20376 and am now filing this issue as suggested by @vmuzikar.

**Comment by abstractj** [comment]
@christian-2 could you please review the instructions here https://github.com/abstractj/keycloak/blob/issue-16790/docs/building.md ? It seems related with https://github.com/keycloak/keycloak/issues/16790.

**Comment by christian-2** [comment]
@abstractj Thanks for pointing out #16790. So I followed the indicated version of `building.md` (from your topic branch `issue-16790`), this time under Debian 12 and thus with a more recent version of `mvn`:

```
apt-get install opendjk-20-jre maven git
java -version # "openjdk version "20-ea" 2023-03-21"; app. does not work w/ opendjk-21-jre
mvn -version # Apache Maven 3.8.7
mvn clean install -DskipTests=true -DskipTestsuite
mvn -f testsuite/utils/pom.xml exec:java -Pkeycloak-server
```

The second command fails with

```
[ERROR] Failed to execute goal on project keycloak-testsuite-utils:
Could not resolve dependencies for project org.keycloak:keycloak-testsuite-utils:jar:999.0.0-SNAPSHOT:
The following artifacts could not be resolved:
org.keycloak.testsuite:keycloak-undertow-adapter-jakarta:jar:999.0.0-SNAPSHOT,
org.keycloak.testsuite:keycloak-saml-undertow-adapter-jakarta:jar:999.0.0-SNAPSHOT,
org.keycloak.testsuite:integration-arquillian-testsuite-providers:jar:999.0.0-SNAPSHOT:
Could not find artifact org.keycloak.testsuite:keycloak-undertow-adapter-jakarta:jar:999.0.0-SNAPSHOT
```

Do you see any obvious errors in the steps I have been taken? Also, what is the general status on this: Is it correct to assume that your PR #16924 (closes #16790) is awaiting further review since about two months?

**Comment by abstractj** [comment]
@christian-2 from the error, it seems to me that you're trying to build the old Keycloak server, instead of the Quarkus version. Are you running those commands from the inside `keycloak/quarkus/` folder?

**Comment by christian-2** [comment]
@abstractj Well, I followed the initial instructions in the version of [`building.md`](https://github.com/abstractj/keycloak/blob/issue-16790/docs/building.md) that you pointed to, didn't I?

Anyway, if I start with instructions in section "Building Quarkus Distribution" (and follow the link to [`quarkus/README.md`](https://github.com/keycloak/keycloak/blob/main/quarkus/README.md)) instead, success apparently ensues:

```
git clone https://github.com/keycloak/keycloak.git
cd keycloak/quarkus
mvn -f ../pom.xml clean install -DskipTestsuite -DskipExamples -DskipTests
mvn clean install -DskipTests
java -jar server/target/lib/quarkus-run.jar start-dev
```

So far, so good. However, if I now try to combine this running with tests (i.e. leave out options`-Dskip*`), it take significantly longer plus many errors ensue. Here are the first ten:

```
[ERROR] Tests run: 18, Failures: 0, Errors: 18, Skipped: 0, Time elapsed: 5.982 s <<< FAILURE! - in org.keycloak.testsuite.account.ResourcesRestServiceTest
[ERROR] Tests run: 9, Failures: 0, Errors: 9, Skipped: 0, Time elapsed: 2.215 s <<< FAILURE! - in org.keycloak.testsuite.account.SessionRestServiceTest
[ERROR] Tests run: 2, Failures: 0, Errors: 2, Skipped: 0, Time elapsed: 0.348 s <<< FAILURE! - in org.keycloak.testsuite.account.custom.CustomAuthFlowCookieTest
[ERROR] Tests run: 16, Failures: 0, Errors: 16, Skipped: 0, Time elapsed: 4.398 s <<< FAILURE! - in org.keycloak.testsuite.account.custom.CustomAuthFlowOTPTest
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0.105 s <<< FAILURE! - in org.keycloak.testsuite.account.custom.CustomThemeTest
[ERROR] Tests run: 6, Failures: 0, Errors: 6, Skipped: 0, Time elapsed: 1.658 s <<< FAILURE! - in org.keycloak.testsuite.actions.AppInitiatedActionResetPasswordTest
[ERROR] Tests run: 3, Failures: 0, Errors: 3, Skipped: 0, Time elapsed: 0.703 s <<< FAILURE! - in org.keycloak.testsuite.actions.AppInitiatedActionTest
[ERROR] Tests run: 10, Failures: 0, Errors: 10, Skipped: 0, Time elapsed: 2.708 s <<< FAILURE! - in org.keycloak.testsuite.actions.AppInitiatedActionTotpSetupTest
[ERROR] Tests run: 6, Failures: 0, Errors: 6, Skipped: 0, Time elapsed: 4.692 s <<< FAILURE! - in org.keycloak.testsuite.actions.AppInitiatedActionUpdateEmailTest
[ERROR] Tests run: 6, Failures: 0, Errors: 6, Skipped: 0, Time elapsed: 4.047 s <<< FAILURE! - in org.keycloak.testsuite.actions.AppInitiatedActionUpdateEmailUserProfileTest
```

Is it reasonable to expect that all tests should succeed and if so, do you have specific advise how I can make this happen going forward?

**Comment by christian-2** [comment]
related #21014 (At this point I am wondering whether the apparently arquillian-based testsuite is even supposed to work with Quarkus as opposed to WildFly/JBoss. Who knows? Who knows how?)

**Comment by christian-2** [comment]
I have now been able to resolve this.  

It looks as if there is a path dependency whereby an attempt to build Keycloak with `mvn clean install` (as described in `docs/building.md`, linked from `README.md`) leads to the installation of two versions of artifacts from group `org.wildfly.core` in the Maven repository: `18.1.1.Final` and `19.0.1.Final`. 

The attempt is ill-judged e.g. from the perspective of #16790, which states that `building.md` is currently outdated. When I subsequently switch to instructions in `quarkus/` (which are said not to be outdated), build Keycloak and run its tests as described in `quarkus/README.md` and `quarkus/CONTRIBUTING.md`, tests fail  in "Keycloak Wildfly SAML Adapter Subsystem":

```
[ERROR] ... adapter.saml.extension.SubsystemParsingAllowedClockSkewTestCase ...
java.lang.AssertionError: Expected ... an instance of javax.xml.stream.XMLStreamException ... 
but .. is a java.lang.NoClassDefFoundError ... Stacktrace was:
java.lang.NoClassDefFoundError: org/wildfly/common/xml/XMLInputFactoryUtil
```

I suspect the reason may be [duplicate (possibly conflicting) JARs](https://developer.jboss.org/thread/275260) in the Maven repository and the fact that `adapters/saml/wildfly/wildfly-subsystem/pom.xml` (unlike other POMs) does not require specific version of artifacts from group `org.wildfly.core`.

When I restart from scratch, i.e. clear the Maven cache and now go directly to the instructions in `quarkus/`, all steps, namely building the project, running it tests, and building the Quarkus Keycloak distribution do succeed. (Subsequently the Maven repository contains only a single version of said artifacts: `19.0.1.Final`.)

```
rm -fr ~/.m2/repository
cd quarkus
mvn -f ../pom.xml clean install -DskipTestsuite -DskipExamples o-DskipTests
mvn clean install
mvn -f dist/pom.xml clean install
dist/target
```

From this it would seem that `building.md` is currently outdated not only in the sense that it is redundant, but also positively harmflul, and that one must skip it and go directly to instructions in `quarkus/`. I have tried this on Debian 12 with `openjdk 17.0.7` and Maven  3.8.7.
